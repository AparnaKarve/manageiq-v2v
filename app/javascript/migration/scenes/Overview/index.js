import React from 'react';
import Equalizer from 'react-equalizer';
import MappingWizardContainer from '../../../components/MappingWizard';
import PlanWizardContainer from '../../../components/PlanWizard';
import { Button, EmptyState, Grid } from 'patternfly-react';
import { bindMethods } from '../../../common/helpers';

class Overview extends React.Component {
  constructor() {
    super();
    this.state = { showMappingWizard: false, showPlanWizard: false };
    bindMethods(this, [
      'getNodes',
      'openMappingWizard',
      'closeMappingWizard',
      'openPlanWizard',
      'closePlanWizard',
      'openPlanWizard'
    ]);
  }
  getNodes(equalizerComponent, equalizerElement) {
    return [this.node1, this.node2];
  }
  openMappingWizard() {
    this.setState({ showMappingWizard: true });
  }
  closeMappingWizard() {
    this.setState({ showMappingWizard: false });
  }
  openPlanWizard() {
    this.setState({ showPlanWizard: true });
  }
  closePlanWizard() {
    this.setState({ showPlanWizard: false });
  }
  render() {
    const { showMappingWizard, showPlanWizard } = this.state;

    const mappingWizard = (
      <MappingWizardContainer
        showWizard={showMappingWizard}
        onHide={this.closeMappingWizard}
      />
    );

    const planWizard = (
      <PlanWizardContainer
        showWizard={showPlanWizard}
        onHide={this.closePlanWizard}
      />
    );

    const overviewCards = (
      <div
        className="row cards-pf"
        ref={n => (this.mainContent = n)}
        style={{ overflow: 'auto', paddingBottom: 100, height: '100%' }}
      >
        <Grid.Col md={12}>
          <Grid.Row>
            <Equalizer nodes={this.getNodes}>
              <Grid.Col xs={6}>
                <div className="spacer" />
                <div className="card-pf">
                  <div className="card-pf-body">
                    <div className="blank-slate-pf" ref={n => (this.node1 = n)}>
                      <EmptyState.Icon />
                      <EmptyState.Title>
                        Infrastructure Mappings
                      </EmptyState.Title>
                      <EmptyState.Info>
                        Create mapping to later be used by a migration plan.
                      </EmptyState.Info>
                      <EmptyState.Action>
                        <Button
                          bsStyle="primary"
                          bsSize="large"
                          onClick={this.openMappingWizard}
                        >
                          Create Infrastructure Mapping
                        </Button>
                      </EmptyState.Action>
                    </div>
                  </div>
                </div>
              </Grid.Col>
              <Grid.Col xs={6}>
                <div className="spacer" />
                <div className="card-pf">
                  <div className="card-pf-body">
                    <div className="blank-slate-pf" ref={n => (this.node2 = n)}>
                      <EmptyState.Icon />
                      <EmptyState.Title>Migration Plans</EmptyState.Title>
                      <EmptyState.Info>
                        Create a migration plan to start migrating.
                      </EmptyState.Info>
                      <EmptyState.Action>
                        <Button
                          bsStyle="primary"
                          bsSize="large"
                          onClick={this.openPlanWizard}
                        >
                          Create Migration Plan
                        </Button>
                      </EmptyState.Action>
                    </div>
                  </div>
                </div>
              </Grid.Col>
            </Equalizer>
          </Grid.Row>

          <Grid.Row>
            <Grid.Col xs={12}>
              <div className="card-pf">
                <div className="card-pf-heading">
                  <h2>0 Migrations in Progress</h2>
                </div>
                <div className="card-pf-body">
                  <p>content</p>
                </div>
              </div>
            </Grid.Col>
          </Grid.Row>

          <Grid.Row>
            <Grid.Col xs={12}>
              <div className="card-pf">
                <div className="card-pf-heading">
                  <h2>0 Completed Migrations</h2>
                </div>
                <div className="card-pf-body">
                  <p>content</p>
                </div>
              </div>
              <div className="spacer" />
            </Grid.Col>
          </Grid.Row>
        </Grid.Col>
      </div>
    );

    return (
      <React.Fragment>
        {overviewCards}
        {showMappingWizard && mappingWizard}
        {showPlanWizard && planWizard}
      </React.Fragment>
    );
  }
}
export default Overview;
