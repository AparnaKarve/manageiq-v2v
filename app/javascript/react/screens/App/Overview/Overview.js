import React from 'react';
import PropTypes from 'prop-types';
import {
  bindMethods,
  Card,
  Breadcrumb,
  CardGrid,
  Spinner,
  Icon,
  Modal,
  Button
} from 'patternfly-react';
import Toolbar from '../../../config/Toolbar';
import * as AggregateCards from './components/AggregateCards';
import InfrastructureMappingsList from './components/InfrastructureMappingsList/InfrastructureMappingsList';
import Migrations from './components/Migrations/Migrations';
import componentRegistry from '../../../../components/componentRegistry';
import getMostRecentRequest from '../common/getMostRecentRequest';

class Overview extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      hasMadeInitialPlansFetch: false
    };

    bindMethods(this, [
      'stopPolling',
      'startPolling',
      'createTransformationPlanRequest',
      'redirectTo',
      'deleteInfraMapping',
      'deleteMessage'
    ]);

    this.mappingWizard = componentRegistry.markup(
      'MappingWizardContainer',
      props.store
    );
    this.planWizard = componentRegistry.markup(
      'PlanWizardContainer',
      props.store
    );
  }

  componentDidMount() {
    const {
      fetchClustersUrl,
      fetchClustersAction,
      fetchTransformationMappingsUrl,
      fetchTransformationMappingsAction,
      fetchTransformationPlansUrl,
      fetchTransformationPlansAction
    } = this.props;

    fetchClustersAction(fetchClustersUrl);
    fetchTransformationMappingsAction(fetchTransformationMappingsUrl);
    fetchTransformationPlansAction(fetchTransformationPlansUrl).then(() => {
      this.setState(() => ({
        hasMadeInitialPlansFetch: true
      }));
      if (!this.pollingInterval) {
        this.startPolling();
      }
    });
  }
  componentWillReceiveProps(nextProps) {
    const {
      isContinuingToPlan,
      fetchTransformationMappingsUrl,
      fetchTransformationMappingsAction,
      fetchTransformationPlansUrl,
      fetchTransformationPlansAction,
      planWizardId,
      continueToPlanAction,
      shouldReloadMappings
    } = this.props;
    const { hasMadeInitialPlansFetch } = this.state;

    if (
      shouldReloadMappings !== nextProps.shouldReloadMappings &&
      nextProps.shouldReloadMappings
    ) {
      // refetech our mappings so that plan wizard has this newly created mapping
      fetchTransformationMappingsAction(fetchTransformationMappingsUrl);
    } else if (
      isContinuingToPlan !== nextProps.isContinuingToPlan &&
      !nextProps.isContinuingToPlan
    ) {
      continueToPlanAction(planWizardId);
    }

    // kill interval if a wizard becomes visble
    if (nextProps.mappingWizardVisible || nextProps.planWizardVisible) {
      this.stopPolling();
    } else if (
      !nextProps.mappingWizardVisible &&
      !nextProps.planWizardVisible &&
      hasMadeInitialPlansFetch &&
      !this.pollingInterval
    ) {
      fetchTransformationPlansAction(fetchTransformationPlansUrl);
      this.startPolling();
    }
  }

  componentDidUpdate(prevProps) {
    const { finishedTransformationPlans, addNotificationAction, yesToDeleteMapping } = this.props;
    const { hasMadeInitialPlansFetch } = this.state;

    if (
      hasMadeInitialPlansFetch &&
      finishedTransformationPlans.length >
        prevProps.finishedTransformationPlans.length
    ) {
      const oldTransformationPlanIds = prevProps.finishedTransformationPlans.map(
        plan => plan.id
      );
      const freshTransformationPlans = finishedTransformationPlans.filter(
        plan => !oldTransformationPlanIds.includes(plan.id)
      );

      freshTransformationPlans.forEach(plan => {
        const mostRecentRequest = getMostRecentRequest(plan.miq_requests);

        let planStatusMessage = sprintf(
          __('%s completed with errors'),
          plan.name
        );
        let planStatus = false;
        if (mostRecentRequest.status === 'Ok') {
          planStatusMessage = sprintf(
            __('%s completed successfully'),
            plan.name
          );
          planStatus = true;
        }

        addNotificationAction({
          message: planStatusMessage,
          notificationType: planStatus ? 'success' : 'error',
          data: {
            id: plan.id
          },
          persistent: !planStatus,
          actionEnabled: true
        });
      });
    }

    if (yesToDeleteMapping) {
      const { deleteInfraMappingAction, mappingToDelete} = this.props;
      deleteInfraMappingAction(mappingToDelete);
    }
  }

  componentWillUnmount() {
    this.stopPolling();
  }

  startPolling() {
    const {
      fetchTransformationPlansAction,
      fetchTransformationPlansUrl
    } = this.props;
    this.pollingInterval = setInterval(() => {
      fetchTransformationPlansAction(fetchTransformationPlansUrl);
    }, 15000);
  }

  stopPolling() {
    if (this.pollingInterval) {
      clearInterval(this.pollingInterval);
      this.pollingInterval = null;
    }
  }

  createTransformationPlanRequest(url, planId) {
    const {
      createTransformationPlanRequestAction,
      fetchTransformationPlansAction,
      fetchTransformationPlansUrl,
      setMigrationsFilterAction,
      retryMigrationAction
    } = this.props;

    createTransformationPlanRequestAction(url).then(() => {
      retryMigrationAction(planId);
      setMigrationsFilterAction('Migration Plans in Progress');
      fetchTransformationPlansAction(fetchTransformationPlansUrl);
    });
  }

  redirectTo(path) {
    const { history } = this.props;
    history.push(path);
  }

  deleteInfraMapping(mapping) {
    const { deleteInfraMappingAction, hideDeleteConfirmationModalAction } = this.props;
    deleteInfraMappingAction(mapping);
    hideDeleteConfirmationModalAction();
  }

  deleteMessage(mappingToDelete) {
    const mappings = [];
    const planNames = [];

    const { transformationPlans, allRequestsWithTasks } = this.props;
    const notStartedPlans = transformationPlans
      .filter(plan => plan.miq_requests.length === 0);

    notStartedPlans.map(plan =>
        mappings.push(
          plan.options.config_info
            .transformation_mapping_id
        )
      );

    notStartedPlans.map(plan =>
        planNames.push(
          plan.name
        )
      );

    const requestsCompletedWithErrors = allRequestsWithTasks
      .filter(request => request.request_state === "finished" && request.status === "Error");

    requestsCompletedWithErrors.map(request =>
        mappings.push(
          request.service_template.options.config_info
            .transformation_mapping_id
        )
      );

    requestsCompletedWithErrors.map(request =>
      planNames.push(
        request.service_template.name
      )
    );

    const mappingWithStyling = <strong>{mappingToDelete.name}</strong>;
    const regularDeleteMessage = sprintf(__('Are you sure you want to delete the infrastructure mapping %s ?'), mappingWithStyling);
    const regularDeleteMessageStyled = <label>{regularDeleteMessage}</label>;

    if (mappings.filter(mapping => mapping === mappingToDelete.id)) {
      const deleteStr = __("The infrastructure mapping is associated with migration plans that include unmigrated VMs. Deleting the mapping will prevent you from migrating the VMs in these plans:");

      const del = <div><h3>{deleteStr}</h3> <br/> <strong>{planNames.map(plan => <ul>{plan}</ul>)}</strong></div>;

      return <div>{del}<br/>{regularDeleteMessageStyled}</div>;

      }
      return regularDeleteMessageStyled;
  }

  render() {
    const {
      showMappingWizardAction,
      showPlanWizardAction,
      mappingWizardVisible,
      planWizardVisible,
      transformationMappings,
      isFetchingTransformationMappings,
      isRejectedTransformationMappings, // eslint-disable-line no-unused-vars
      transformationPlans,
      allRequestsWithTasks,
      reloadCard,
      isFetchingAllRequestsWithTasks,
      requestsWithTasksPreviouslyFetched,
      notStartedTransformationPlans,
      activeTransformationPlans,
      finishedTransformationPlans,
      isCreatingTransformationPlanRequest,
      clusters,
      migrationsFilter,
      setMigrationsFilterAction,
      deleteConfirmationModal,
      mappingToDelete,
      deleteInfraMappingAction,
      setMappingToDelete,
      showDeleteConfirmationModalAction,
      hideDeleteConfirmationModalAction,
      yesAndHideDeleteConfirmationModalAction,
      yesToDeleteMapping
    } = this.props;

    const transformationMappingsUsedInProgressRequests = () => {
      // const inProgressRequests = allRequestsWithTasks.filter(request => request.fulfilled_on === null);
      const mappings = [];

      allRequestsWithTasks
        .filter(request => request.fulfilled_on === null)
        .map(request =>
          mappings.push(
            request.service_template.options.config_info
              .transformation_mapping_id
          )
        );
      return mappings;
    };

    const transformationMappingsWithNoPlanAssociations = () => {
      const mappings = [];

      transformationMappings
        .filter(mapping => mapping.service_template.length === 0)
        .map(mapping =>
          mappings.push(
            mapping.id
          )
        );
      return mappings;
    };

    const transformationMappingsWithRequestsCompletedSuccessfully = () => {
      const mappings = [];

      allRequestsWithTasks
        .filter(request => request.request_state === "finished" && request.status === "Ok")
        .map(request =>
          mappings.push(
            request.service_template.options.config_info
              .transformation_mapping_id
          )
        );
      return mappings;
    };

    const transformationMappingsWithNotStartedPlans = () => {
      const mappings = [];

      transformationPlans
        .filter(plan => plan.miq_requests.length === 0)
        .map(plan =>
          mappings.push(
            plan.options.config_info
              .transformation_mapping_id
          )
        );
      return mappings;
    };

    const transformationMappingsWithRequestsCompletedWithErrors = () => {
      const mappings = [];

      allRequestsWithTasks
        .filter(request => request.request_state === "finished" && request.status === "Error")
        .map(request =>
          mappings.push(
            request.service_template.options.config_info
              .transformation_mapping_id
          )
        );
      return mappings;
    };

    const aggregateDataCards = (
      <div className="row-cards-pf">
        <Card.HeightMatching selector={['.card-pf-match-height']}>
          <CardGrid.Col xs={6} sm={3}>
            <AggregateCards.NotStartedTransformationPlans
              notStartedPlans={notStartedTransformationPlans}
              loading={
                isFetchingAllRequestsWithTasks &&
                !requestsWithTasksPreviouslyFetched
              }
            />
          </CardGrid.Col>
          <CardGrid.Col xs={6} sm={3}>
            <AggregateCards.ActiveTransformationPlans
              activePlans={activeTransformationPlans}
              allRequestsWithTasks={allRequestsWithTasks}
              reloadCard={reloadCard}
              loading={
                isFetchingAllRequestsWithTasks &&
                !requestsWithTasksPreviouslyFetched
              }
            />
          </CardGrid.Col>
          <CardGrid.Col xs={6} sm={3}>
            <AggregateCards.FinishedTransformationPlans
              finishedPlans={finishedTransformationPlans}
              loading={
                isFetchingAllRequestsWithTasks &&
                !requestsWithTasksPreviouslyFetched
              }
            />
          </CardGrid.Col>
          <CardGrid.Col xs={6} sm={3}>
            <AggregateCards.InfrastructureMappings
              mappings={transformationMappings}
              loading={isFetchingTransformationMappings}
            />
          </CardGrid.Col>
        </Card.HeightMatching>
      </div>
    );

    const overviewContent = (
      <div
        className="row cards-pf"
        style={{ overflow: 'auto', paddingBottom: 1, height: '100%' }}
      >
        {aggregateDataCards}
        <Spinner
          loading={
            isFetchingTransformationMappings ||
            (isFetchingAllRequestsWithTasks &&
              !requestsWithTasksPreviouslyFetched)
          }
          style={{ marginTop: 200 }}
        >
          {transformationMappings.length > 0 && (
            <Migrations
              activeFilter={migrationsFilter}
              setActiveFilter={setMigrationsFilterAction}
              transformationPlans={transformationPlans}
              allRequestsWithTasks={allRequestsWithTasks}
              reloadCard={reloadCard}
              notStartedPlans={notStartedTransformationPlans}
              activeTransformationPlans={activeTransformationPlans}
              finishedTransformationPlans={finishedTransformationPlans}
              createMigrationPlanClick={showPlanWizardAction}
              createTransformationPlanRequestClick={
                this.createTransformationPlanRequest
              }
              isCreatingTransformationPlanRequest={
                isCreatingTransformationPlanRequest
              }
              redirectTo={this.redirectTo}
            />
          )}
          <InfrastructureMappingsList
            clusters={clusters}
            transformationMappings={transformationMappings}
            createInfraMappingClick={showMappingWizardAction}
            requestsInProgressMappings={transformationMappingsUsedInProgressRequests()}
            // noPlanAssociationsMappings={transformationMappingsWithNoPlanAssociations()}
            // requestsCompletedSuccessfullyMappings={transformationMappingsWithRequestsCompletedSuccessfully()}
            // notStartedPlansMappings={transformationMappingsWithNotStartedPlans()}
            // requestsCompletedWithErrorsMappings={transformationMappingsWithRequestsCompletedWithErrors()}
            setMappingToDelete={setMappingToDelete}
            showDeleteConfirmationModalAction={showDeleteConfirmationModalAction}
          />
          <Modal
            show={deleteConfirmationModal}
            onHide={hideDeleteConfirmationModalAction}
          >
            <Modal.Header>
              <Modal.CloseButton onClick={hideDeleteConfirmationModalAction} />
              <Modal.Title>{__('Delete Infrastructure Mapping')}</Modal.Title>
            </Modal.Header>
            <Modal.Body className="warning-modal-body">
              <div className="warning-modal-body--icon">
                <Icon type="pf" name="delete" />
              </div>
              <div className="warning-modal-body--list">
                <h3>
                  {/*{mappingToDelete && sprintf(__('Are you sure you want to delete the infrastructure mapping "%s"?'), mappingToDelete.name)}*/}
                  {mappingToDelete && this.deleteMessage(mappingToDelete)}
                </h3>
              </div>
            </Modal.Body>
            <Modal.Footer>
              <Button
                bsStyle="default"
                className="btn-cancel"
                onClick={hideDeleteConfirmationModalAction}
              >
                {__("Cancel")}
              </Button>
              <Button bsStyle="primary" onClick={yesAndHideDeleteConfirmationModalAction}>
                {__("Delete")}
              </Button>
            </Modal.Footer>
          </Modal>
        </Spinner>
      </div>
    );

    const toolbarContent = (
      <Toolbar>
        <Breadcrumb.Item href="/dashboard/maintab?tab=compute">
          Compute
        </Breadcrumb.Item>
        <Breadcrumb.Item active>Migration</Breadcrumb.Item>
      </Toolbar>
    );

    return (
      <React.Fragment>
        {toolbarContent}
        {overviewContent}
        {mappingWizardVisible && this.mappingWizard}
        {planWizardVisible && this.planWizard}
      </React.Fragment>
    );
  }
}
Overview.propTypes = {
  store: PropTypes.object,
  showMappingWizardAction: PropTypes.func,
  showPlanWizardAction: PropTypes.func,
  addNotificationAction: PropTypes.func,
  mappingWizardVisible: PropTypes.bool,
  planWizardVisible: PropTypes.bool,
  transformationPlans: PropTypes.array,
  allRequestsWithTasks: PropTypes.array,
  reloadCard: PropTypes.bool,
  fetchTransformationPlansUrl: PropTypes.string,
  fetchTransformationPlansAction: PropTypes.func,
  isFetchingAllRequestsWithTasks: PropTypes.bool,
  requestsWithTasksPreviouslyFetched: PropTypes.bool,
  notStartedTransformationPlans: PropTypes.array,
  activeTransformationPlans: PropTypes.array,
  finishedTransformationPlans: PropTypes.array,
  transformationMappings: PropTypes.array,
  fetchTransformationMappingsUrl: PropTypes.string,
  fetchTransformationMappingsAction: PropTypes.func,
  isFetchingTransformationMappings: PropTypes.bool,
  isRejectedTransformationMappings: PropTypes.bool,
  createTransformationPlanRequestAction: PropTypes.func,
  isCreatingTransformationPlanRequest: PropTypes.string,
  isContinuingToPlan: PropTypes.bool,
  planWizardId: PropTypes.string,
  continueToPlanAction: PropTypes.func,
  shouldReloadMappings: PropTypes.bool,
  fetchClustersAction: PropTypes.func,
  fetchClustersUrl: PropTypes.string,
  clusters: PropTypes.array,
  migrationsFilter: PropTypes.string,
  setMigrationsFilterAction: PropTypes.func,
  retryMigrationAction: PropTypes.func,
  history: PropTypes.object,
  deleteInfraMappingAction: PropTypes.func,
  deleteConfirmationModal: PropTypes.bool,
  mappingToDelete: PropTypes.object,
  setMappingToDelete: PropTypes.func,
  showDeleteConfirmationModalAction: PropTypes.func,
  hideDeleteConfirmationModalAction: PropTypes.func,
  yesAndHideDeleteConfirmationModalAction: PropTypes.func,
  yesToDeleteMapping: PropTypes.bool
};
export default Overview;
