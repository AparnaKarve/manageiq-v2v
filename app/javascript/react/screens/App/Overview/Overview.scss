@import './screens/MappingWizard/MappingWizard.scss';
@import './screens/PlanWizard/components/PlanWizardResultsStep/PlanWizardResultsStep.scss';

.card-pf-body .blank-slate-pf {
  background-color: #fff;
}

@media (min-width: 768px) {
  .card-pf-body .blank-slate-pf {
    padding: 20px 20px;
  }
}

.csv-upload-dropzone {
  border: 3px solid #fff;
}

.csv-upload-dropzone.active {
  border: 3px dashed #9b9b9b; // Matches upload icon color
}
