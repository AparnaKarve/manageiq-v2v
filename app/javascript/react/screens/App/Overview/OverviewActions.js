import URI from 'urijs';
import API from '../../../../common/API';

import {
  SHOW_MAPPING_WIZARD,
  SHOW_PLAN_WIZARD,
  HIDE_MAPPING_WIZARD,
  FETCH_V2V_TRANSFORMATION_MAPPINGS,
  FETCH_V2V_TRANSFORMATION_PLANS,
  FETCH_V2V_ALL_REQUESTS_WITH_TASKS,
  CREATE_V2V_TRANSFORMATION_PLAN_REQUEST,
  V2V_FETCH_CLUSTERS,
  V2V_SET_MIGRATIONS_FILTER,
  V2V_RETRY_MIGRATION,
  SHOW_DELETE_CONFIRMATION_MODAL,
  HIDE_DELETE_CONFIRMATION_MODAL,
  YES_AND_HIDE_DELETE_CONFIRMATION_MODAL,
  MAPPING_TO_BE_DELETED,
  DELETE_INFRA_MAPPING
} from './OverviewConstants';

export const showMappingWizardAction = () => dispatch => {
  dispatch({
    type: SHOW_MAPPING_WIZARD
  });
};

export const showPlanWizardAction = id => dispatch => {
  dispatch({
    type: SHOW_PLAN_WIZARD,
    payload: id
  });
};

const _createTransformationPlanRequestActionCreator = url => dispatch =>
  dispatch({
    type: CREATE_V2V_TRANSFORMATION_PLAN_REQUEST,
    payload: {
      promise: API.post(url, { action: 'order' }),
      data: url
    }
  });

export const createTransformationPlanRequestAction = url => {
  const uri = new URI(url);
  return _createTransformationPlanRequestActionCreator(uri.toString());
};

const _getTransformationMappingsActionCreator = url => dispatch =>
  dispatch({
    type: FETCH_V2V_TRANSFORMATION_MAPPINGS,
    payload: API.get(url)
  });

export const fetchTransformationMappingsAction = url => {
  const uri = new URI(url);
  return _getTransformationMappingsActionCreator(uri.toString());
};

const fetchTasksForAllRequests = (allRequests, dispatch) => {
  if (allRequests.length > 0) {
    dispatch({
      type: FETCH_V2V_ALL_REQUESTS_WITH_TASKS,
      payload: new Promise((resolve, reject) => {
        API.post(
          '/api/requests?expand=resource&attributes=miq_request_tasks,service_template',
          {
            action: 'query',
            resources: allRequests
          }
        )
          .then(responseRequestsWithTasks => {
            resolve(responseRequestsWithTasks);
          })
          .catch(e => reject(e));
      })
    });
  }
};

const collectAllRequests = plan =>
  plan.miq_requests.map(request => Object.assign({}, { href: request.href }));

const _getTransformationPlansActionCreator = url => dispatch =>
  dispatch({
    type: FETCH_V2V_TRANSFORMATION_PLANS,
    payload: new Promise((resolve, reject) => {
      API.get(url)
        .then(response => {
          resolve(response);
          const allPlansWithRequests = response.data.resources;

          const allRequests = [];
          const mergedRequests = [].concat(
            ...allRequests.concat(
              allPlansWithRequests.map(plan => collectAllRequests(plan))
            )
          );

          fetchTasksForAllRequests(mergedRequests, dispatch);
        })
        .catch(e => reject(e));
    })
  });

export const fetchTransformationPlansAction = url => {
  const uri = new URI(url);
  return _getTransformationPlansActionCreator(uri.toString());
};

export const continueToPlanAction = id => dispatch => {
  dispatch({
    type: HIDE_MAPPING_WIZARD
  });
  dispatch({
    type: SHOW_PLAN_WIZARD,
    payload: { id }
  });
};

const _getClustersActionCreator = url => dispatch =>
  dispatch({
    type: `${V2V_FETCH_CLUSTERS}`,
    payload: API.get(url)
  });

export const fetchClustersAction = url => {
  const uri = new URI(url);
  return _getClustersActionCreator(uri.toString());
};

export const setMigrationsFilterAction = filter => ({
  type: V2V_SET_MIGRATIONS_FILTER,
  payload: filter
});

export const retryMigrationAction = planId => ({
  type: V2V_RETRY_MIGRATION,
  payload: planId
});

export const hideDeleteConfirmationModalAction = () => dispatch => {
  dispatch({
    type: HIDE_DELETE_CONFIRMATION_MODAL,
    payload: false
  });
};

export const yesAndHideDeleteConfirmationModalAction = () => dispatch => {
  dispatch({
    type: YES_AND_HIDE_DELETE_CONFIRMATION_MODAL
  });
};

export const showDeleteConfirmationModalAction = mapping => dispatch => {
  dispatch({
    type: SHOW_DELETE_CONFIRMATION_MODAL,
    payload: true
  });
};

export const setMappingToDelete = mapping => dispatch => {
  dispatch({
    type: MAPPING_TO_BE_DELETED,
    payload: mapping
  });
};

export const deleteInfraMappingAction = mapping => dispatch => {
  dispatch({
    type: DELETE_INFRA_MAPPING,
    payload: new Promise((resolve, reject) => {
      API.post(`/api/transformation_mappings/${mapping.id}`, {
        action: 'delete'
      })
        .then(response => {
          resolve(response);
        })
        .catch(e => reject(e));
    })
  });
};
