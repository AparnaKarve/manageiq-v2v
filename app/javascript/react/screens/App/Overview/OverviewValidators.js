import { validateSchema } from '../../../../data/schemaHelpers';
import { clustersSchema } from '../../../../data/models/clusters';
import { datastoresSchema } from '../../../../data/models/datastores';
import { networksSchema } from '../../../../data/models/networks';
import { plansSchema } from '../../../../data/models/plans';
import { providersSchema } from '../../../../data/models/provider';
import { requestsSchema } from '../../../../data/models/requests';
import { mappingsSchema } from '../../../../data/models/transformationMappings';
import { serviceTemplatePlaybookSchema } from '../../../../data/models/playbooks';

export const validateOverviewClusters = clusters => {
  validateSchema(clustersSchema, clusters);
};

export const validateOverviewDatastores = datastores => {
  validateSchema(datastoresSchema, datastores);
};

export const validateOverviewNetworks = networks => {
  validateSchema(networksSchema, networks);
};

export const validateOverviewPlans = plans => {
  validateSchema(plansSchema, plans);
};

export const validateOverviewProviders = providers => {
  validateSchema(providersSchema, providers);
};

export const validateOverviewRequests = requests => {
  validateSchema(requestsSchema, requests);
};

export const validateOverviewMappings = mappings => {
  validateSchema(mappingsSchema, mappings);
};

export const validateServiceTemplatePlaybooks = playbooks => {
  validateSchema(serviceTemplatePlaybookSchema, playbooks);
};
