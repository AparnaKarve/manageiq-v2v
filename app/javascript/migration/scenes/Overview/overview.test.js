import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Overview from './index';

describe('Overview component', () => {
  it('renders the overview', () => {
    const wrapper = mount(<Overview />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
