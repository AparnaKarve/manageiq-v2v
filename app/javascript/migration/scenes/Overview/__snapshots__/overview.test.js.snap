// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview component renders the overview 1`] = `
<Overview>
  <div
    className="row cards-pf"
    style={
      Object {
        "height": "100%",
        "overflow": "auto",
        "paddingBottom": 100,
      }
    }
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <div
        className="col-md-12"
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <div
            className="row"
          >
            <Equalizer
              byRow={true}
              enabled={[Function]}
              nodes={[Function]}
              property="height"
            >
              <div
                onLoad={[Function]}
              >
                <Col
                  bsClass="col"
                  componentClass="div"
                  xs={6}
                >
                  <div
                    className="col-xs-6"
                  >
                    <div
                      className="spacer"
                    />
                    <div
                      className="card-pf"
                    >
                      <div
                        className="card-pf-body"
                      >
                        <div
                          className="blank-slate-pf"
                        >
                          <EmptyStateIcon
                            name="add-circle-o"
                            type="pf"
                          >
                            <div
                              className="blank-slate-pf-icon"
                            >
                              <Icon
                                name="add-circle-o"
                                type="pf"
                              >
                                <PatternflyIcon
                                  name="add-circle-o"
                                >
                                  <span
                                    aria-hidden="true"
                                    className="pficon pficon-add-circle-o"
                                  />
                                </PatternflyIcon>
                              </Icon>
                            </div>
                          </EmptyStateIcon>
                          <EmptyStateTitle>
                            <h4
                              className="h1 blank-slate-pf-title"
                            >
                              Infrastructure Mappings
                            </h4>
                          </EmptyStateTitle>
                          <EmptyStateInfo>
                            <p
                              className="blank-slate-pf-info"
                            >
                              Create mapping to later be used by a migration plan.
                            </p>
                          </EmptyStateInfo>
                          <EmptyStateAction>
                            <div
                              className="blank-slate-pf-main-action"
                            >
                              <Button
                                bsSize="large"
                                bsStyle="primary"
                                onClick={[Function]}
                              >
                                <Button
                                  active={false}
                                  block={false}
                                  bsClass="btn"
                                  bsSize="large"
                                  bsStyle="primary"
                                  disabled={false}
                                  onClick={[Function]}
                                >
                                  <button
                                    className="btn btn-lg btn-primary"
                                    disabled={false}
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Create Infrastructure Mapping
                                  </button>
                                </Button>
                              </Button>
                            </div>
                          </EmptyStateAction>
                        </div>
                      </div>
                    </div>
                  </div>
                </Col>
                <Col
                  bsClass="col"
                  componentClass="div"
                  xs={6}
                >
                  <div
                    className="col-xs-6"
                  >
                    <div
                      className="spacer"
                    />
                    <div
                      className="card-pf"
                    >
                      <div
                        className="card-pf-body"
                      >
                        <div
                          className="blank-slate-pf"
                        >
                          <EmptyStateIcon
                            name="add-circle-o"
                            type="pf"
                          >
                            <div
                              className="blank-slate-pf-icon"
                            >
                              <Icon
                                name="add-circle-o"
                                type="pf"
                              >
                                <PatternflyIcon
                                  name="add-circle-o"
                                >
                                  <span
                                    aria-hidden="true"
                                    className="pficon pficon-add-circle-o"
                                  />
                                </PatternflyIcon>
                              </Icon>
                            </div>
                          </EmptyStateIcon>
                          <EmptyStateTitle>
                            <h4
                              className="h1 blank-slate-pf-title"
                            >
                              Migration Plans
                            </h4>
                          </EmptyStateTitle>
                          <EmptyStateInfo>
                            <p
                              className="blank-slate-pf-info"
                            >
                              Create a migration plan to start migrating.
                            </p>
                          </EmptyStateInfo>
                          <EmptyStateAction>
                            <div
                              className="blank-slate-pf-main-action"
                            >
                              <Button
                                bsSize="large"
                                bsStyle="primary"
                                onClick={[Function]}
                              >
                                <Button
                                  active={false}
                                  block={false}
                                  bsClass="btn"
                                  bsSize="large"
                                  bsStyle="primary"
                                  disabled={false}
                                  onClick={[Function]}
                                >
                                  <button
                                    className="btn btn-lg btn-primary"
                                    disabled={false}
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Create Migration Plan
                                  </button>
                                </Button>
                              </Button>
                            </div>
                          </EmptyStateAction>
                        </div>
                      </div>
                    </div>
                  </div>
                </Col>
              </div>
            </Equalizer>
          </div>
        </Row>
        <Row
          bsClass="row"
          componentClass="div"
        >
          <div
            className="row"
          >
            <Col
              bsClass="col"
              componentClass="div"
              xs={12}
            >
              <div
                className="col-xs-12"
              >
                <div
                  className="card-pf"
                >
                  <div
                    className="card-pf-heading"
                  >
                    <h2>
                      0 Migrations in Progress
                    </h2>
                  </div>
                  <div
                    className="card-pf-body"
                  >
                    <p>
                      content
                    </p>
                  </div>
                </div>
              </div>
            </Col>
          </div>
        </Row>
        <Row
          bsClass="row"
          componentClass="div"
        >
          <div
            className="row"
          >
            <Col
              bsClass="col"
              componentClass="div"
              xs={12}
            >
              <div
                className="col-xs-12"
              >
                <div
                  className="card-pf"
                >
                  <div
                    className="card-pf-heading"
                  >
                    <h2>
                      0 Completed Migrations
                    </h2>
                  </div>
                  <div
                    className="card-pf-body"
                  >
                    <p>
                      content
                    </p>
                  </div>
                </div>
                <div
                  className="spacer"
                />
              </div>
            </Col>
          </div>
        </Row>
      </div>
    </Col>
  </div>
</Overview>
`;
